### PR TITLE
Replace use of #ifdef by template parameter in symex_assign

### PR DESCRIPTION
**Pull request by romainbrenguier**
This make sure both branches could be compile (which was no longer the case), and open the possibility to unit-test both branches. 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 43fffdf2e644b35adef592ec6e95ac3bb1530985).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116945274
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by smowton — APPROVED** [review]

File: src/goto-symex/symex_assign.cpp [hunk 1]
```diff
@@ -269,31 +291,38 @@ static assignmentt shift_indexed_access_to_lhs(
...
+          ssa_rhs = [&]() -> exprt {
+            if(use_update)
+            {
+              update_exprt new_rhs{idx.array(), exprt{}, ssa_rhs};
+              new_rhs.designator().push_back(index_designatort{idx.index()});
+              return std::move(new_rhs);
+            }
+            else
+              return with_exprt{idx.array(), idx.index(), ssa_rhs};
+          }();
```
Comment (tautschnig) [inline src/goto-symex/symex_assign.cpp:303]: I don't get the point of this lambda function?!

File: src/goto-symex/symex_assign.cpp [hunk 2]
```diff
@@ -269,31 +291,38 @@ static assignmentt shift_indexed_access_to_lhs(
...
+              return std::move(new_rhs);
+            }
+            else
+            {
+              with_exprt new_rhs(
+                member.compound(), exprt(ID_member_name), ssa_rhs);
+              new_rhs.where().set(ID_component_name, component_name);
+              return std::move(new_rhs);
+            }
+          }();
```
Comment (tautschnig) [inline src/goto-symex/symex_assign.cpp:325]: Again, no idea why a lambda is the right thing here?

**Review by tautschnig — APPROVED** [review]

File: src/goto-symex/symex_assign.cpp (hunk 1 above)
Comment (romainbrenguier) [reply src/goto-symex/symex_assign.cpp:303]: This is to make clear that this part of the code is just to find the new value of ssa_rhs. The alternative is to have directly an `if` statement and we wouldn't know what it's mutating without looking at the code inside the branches.

**Pull request merged**